pkg/ccl/backupccl/backup_span_maker.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package backupccl
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/rangedesc"
+	"github.com/cockroachdb/errors"
+)
+
+type requestSpanMaker struct {
+	requestSpans     []spanAndTime
+	specs            *execinfrapb.BackupDataSpec
+	rangeSizedSpans  bool
+	rangeIterFactory rangedesc.IteratorFactory
+}
+
+const maxChunkSize = 100
+
+// Aim to make at least 4 chunks per worker, ensuring size is >=1 and <= max.
+const softMinPerWorkerChunkCount = 4
+
+func (rsm *requestSpanMaker) queueWork(numSenders int) chan []spanAndTime {
+	todo := make(chan []spanAndTime, len(rsm.requestSpans))
+
+	chunkSize := (len(rsm.requestSpans) / (numSenders * softMinPerWorkerChunkCount)) + 1
+	if chunkSize > maxChunkSize {
+		chunkSize = maxChunkSize
+	}
+
+	chunk := make([]spanAndTime, 0, chunkSize)
+	for i := range rsm.requestSpans {
+		if !rsm.rangeSizedSpans {
+			todo <- []spanAndTime{rsm.requestSpans[i]}
+			continue
+		}
+
+		chunk = append(chunk, rsm.requestSpans[i])
+		if len(chunk) > chunkSize {
+			todo <- chunk
+			chunk = make([]spanAndTime, 0, chunkSize)
+		}
+	}
+	if len(chunk) > 0 {
+		todo <- chunk
+	}
+	return todo
+}
+
+func (rsm *requestSpanMaker) spanCountFromSpecs() int {
+	return len(rsm.specs.Spans) + len(rsm.specs.IntroducedSpans)
+}
+
+func initRequestSpanMaker(
```

Comment:
Having read this and the usage in the processor, I'm not sure I really have a good intuitive sense for why the API boundaries are where they are: why do I pass `spec` to init, but then nothing to `addSpans` ? why am I not passing the spans to addSpans, or why doesn't init just add the spans? 

This almost feels like we don't need a (stateful?) object at all, but maybe just a couple free functions that compose together? and the call in the processor looks like it could be just a single function call, eg `todo, err := fn(ctx, rdi, spec, workerCount, settings)` ?

Reply:
yeah, that's a better design, given that we call all of these functions once. My initial approach was to keep the same call sites and reduce the size of the function signature, but I went with your suggestion.